File: shared/src/main/java/org/tiqr/authenticator/security/Secret.java
package org.tiqr.authenticator.security;

import android.content.Context;
import android.util.Log;

import org.tiqr.authenticator.datamodel.Identity;
import org.tiqr.authenticator.exceptions.SecurityFeaturesException;

import java.io.IOException;
import java.security.InvalidKeyException;
import java.security.KeyStoreException;
import java.security.NoSuchAlgorithmException;
import java.security.UnrecoverableEntryException;
import java.security.UnrecoverableKeyException;
import java.security.cert.CertificateException;

import javax.crypto.SecretKey;
import javax.crypto.spec.SecretKeySpec;

public class Secret {

    public enum Type {
        PINCODE,
        FINGERPRINT
    }

    private final static String FINGERPRINT_SUFFIX = "org.tiqr.FP";

    private Identity _identity = null;
    private SecretKey _secret = null;
    private SecretStore _store = null;
    private Context _ctx = null;

    public static Secret secretForIdentity(Identity identity, Context context) throws KeyStoreException {
        return new Secret(identity, context);
    }

    private Secret(Identity identity, Context context) throws KeyStoreException {
        _identity = identity;
        _store = new SecretStore(context);
        _ctx = context;
    }

    public SecretKey getSecret(SecretKey sessionKey, Type type) throws InvalidKeyException, SecurityFeaturesException, CertificateException, UnrecoverableEntryException, NoSuchAlgorithmException, KeyStoreException, IOException {
        if (_secret == null) {
            _loadFromKeyStore(sessionKey, type);
        }
        return _secret;
    }

    public void setSecret(SecretKey secret) {
        _secret = secret;
    }

    private String getId(Type type) {
        if (type == Type.PINCODE) {
            return Long.toString(_identity.getId());
        } else {
            return Long.toString(_identity.getId()) + FINGERPRINT_SUFFIX;
        }
    }

    private SecretKey _loadFromKeyStore(SecretKey sessionKey, Type type) throws SecurityFeaturesException, InvalidKeyException, CertificateException,
            NoSuchAlgorithmException, KeyStoreException, IOException, UnrecoverableEntryException {
        SecretKey deviceKey = Encryption.getDeviceKey(_ctx);
        CipherPayload civ = _store.getSecretKey(getId(type), deviceKey);
        if (civ == null || civ.cipherText == null) {
            throw new InvalidKeyException("Requested key not found.");
        }
        _secret = new SecretKeySpec(Encryption.decrypt(civ, sessionKey), "RAW");
        if (civ.iv == null) {
            // Old keys didn't store the iv, so upgrade it to a new key.
            Log.i("encryption", "Found old style key; upgrading to new key.");
            storeInKeyStore(sessionKey, type);
        }
        return _secret;
    }

    public void storeInKeyStore(SecretKey sessionKey, Type type) throws SecurityFeaturesException, CertificateException, NoSuchAlgorithmException, KeyStoreException, IOException {
        CipherPayload civ = Encryption.encrypt(_secret.getEncoded(), sessionKey);
        _store.setSecretKey(getId(type), civ, Encryption.getDeviceKey(_ctx));
    }
}

File: shared/src/main/java/org/tiqr/authenticator/security/Secret.kt
package org.tiqr.authenticator.security

import android.content.Context
import android.util.Log

import org.tiqr.authenticator.datamodel.Identity
import org.tiqr.authenticator.exceptions.SecurityFeaturesException

import java.io.IOException
import java.security.InvalidKeyException
import java.security.KeyStoreException
import java.security.NoSuchAlgorithmException
import java.security.UnrecoverableEntryException
import java.security.cert.CertificateException

import javax.crypto.SecretKey
import javax.crypto.spec.SecretKeySpec

class Secret @Throws(KeyStoreException::class)
private constructor(val identity: Identity, val context: Context) {

    private var secret: SecretKey? = null
    private val store = SecretStore(context)

    enum class Type {
        PINCODE,
        FINGERPRINT
    }

    @Throws(InvalidKeyException::class, SecurityFeaturesException::class, CertificateException::class, UnrecoverableEntryException::class, NoSuchAlgorithmException::class, KeyStoreException::class, IOException::class)
    fun getSecret(sessionKey: SecretKey, type: Type): SecretKey? {
        if (secret == null) {
            loadFromKeyStore(sessionKey, type)
        }
        return secret
    }

    fun setSecret(secret: SecretKey) {
        this.secret = secret
    }

    private fun getId(type: Type): String {
        return if (type == Type.PINCODE) {
            "${identity.id}"
        } else {
            "${identity.id}$FINGERPRINT_SUFFIX"
        }
    }

    @Throws(SecurityFeaturesException::class, InvalidKeyException::class, CertificateException::class, NoSuchAlgorithmException::class, KeyStoreException::class, IOException::class, UnrecoverableEntryException::class)
    private fun loadFromKeyStore(sessionKey: SecretKey, type: Type): SecretKey? {
        val deviceKey = Encryption.getDeviceKey(context)
        val civ = store.getSecretKey(getId(type), deviceKey)

        secret = SecretKeySpec(Encryption.decrypt(civ, sessionKey), "RAW")

        if (civ.iv == null) {
            // Old keys didn't store the iv, so upgrade it to a new key.
            Log.i("encryption", "Found old style key; upgrading to new key.")
            storeInKeyStore(sessionKey, type)
        }
        return secret
    }

    @Throws(SecurityFeaturesException::class, CertificateException::class, NoSuchAlgorithmException::class, KeyStoreException::class, IOException::class)
    fun storeInKeyStore(sessionKey: SecretKey, type: Type) {
        val civ = Encryption.encrypt(secret!!.encoded, sessionKey)
        store.setSecretKey(getId(type), civ, Encryption.getDeviceKey(context))
    }

    companion object {

        private val FINGERPRINT_SUFFIX = "org.tiqr.FP"

        @Throws(KeyStoreException::class)
        fun secretForIdentity(identity: Identity, context: Context): Secret {
            return Secret(identity, context)
        }
    }
}
